jaseci_ai_kit/jaseci_ai_kit/modules/tts/tts.py

```diff
@@ -0,0 +1,169 @@
+
+    Parameters:
+    -----------
+    state_dict: Dictionary
+
+    Return:
+    -----------
+    ret: Boolean
+
+    """
+    ret = False
+    for key, _ in state_dict.items():
+        if key.find("module.") != -1:
+            ret = True
+            break
+    return ret
+
+
+def _unwrap_distributed(state_dict):
+    """
+    Unwraps model from DistributedDataParallel.
+    DDP wraps model in additional "module.", it needs to be removed for single
+    GPU inference.
+
+    Parameters:
+    -----------
+
+    state_dict: Dictionary, model's state dict
+
+    Return:
+    -----------
+
+    new_state_dict = Dictionary, modified model's state dict
+    """
+    new_state_dict = {}
+    for key, value in state_dict.items():
+        new_key = key.replace("module.1.", "")
+        new_key = new_key.replace("module.", "")
+        new_state_dict[new_key] = value
+    return new_state_dict
+
+
+def load_tacotron(checkpoint, force_reload):
+    """
+    Download tacotron2 checkpoints and loading state dictionery
+
+    Parameters:
+    -----------
+    checkpoint: checkpoint file
+
+    Return:
+    -----------
+    tacotron2 = Model
+    """
+    ckpt_file = _download_checkpoint(checkpoint, force_reload)
+    ckpt = torch.load(ckpt_file, map_location=torch.device("cpu"))
+    state_dict = ckpt["state_dict"]
+    if _checkpoint_from_distributed(state_dict):
+        state_dict = _unwrap_distributed(state_dict)
+    config = ckpt["config"]
+    tacotron2 = Tacotron2.Tacotron2(**config)
+    tacotron2.load_state_dict(state_dict)
+
+    return tacotron2
+
+
+def load_waveglow(checkpoint, force_reload):
+
+    ckpt_file = _download_checkpoint(checkpoint, force_reload)
+    ckpt = torch.load(ckpt_file, map_location=torch.device("cpu"))
+    state_dict = ckpt["state_dict"]
+    if _checkpoint_from_distributed(state_dict):
+        state_dict = _unwrap_distributed(state_dict)
+    config = ckpt["config"]
+    waveglow = Waveglow.WaveGlow(**config)
+    waveglow.load_state_dict(state_dict)
+
+    return waveglow
+
+
+def make_utils(force_reload):
+
+    tacotron2 = load_tacotron(taco_checkpoint, force_reload)
+    waveglow = load_waveglow(waglo_checkpoint, force_reload)
+    waveglow = waveglow.remove_weightnorm(waveglow)
+    utils = torch.hub.load("NVIDIA/DeepLearningExamples:torchhub", "nvidia_tts_utils")
+
+    return tacotron2, waveglow, utils
+
+
+tacotron2, waveglow, utils = make_utils(force_reload)
+
+
+def prediction(input_text, model=tacotron2, vocorder=waveglow, utils=utils):
+
+    sequences, lengths = utils.prepare_input_sequence([input_text], cpu_run=True)
+    with torch.no_grad():
+        mel, _, _ = tacotron2.infer(sequences, lengths)
+        audio = waveglow.infer(mel)
+    audio_numpy = audio[0].data.cpu().numpy()
+    rate = 22050
+
+    write("audio.wav", rate, audio_numpy)
+
+    return audio_numpy
+
+
+@jaseci_action(act_group=["synthesize"], allow_remote=True)
+def synthesize(text: str, threshold: float = 0.7):
```

Comment:
* instructions on how to use this end-to-end in a workflow, e.g. how a frontend can use the response here to construct an audio file.
* return a list instead of numpy array because np array does not work in Jac.
* parameter on whether to convert the return to base64
* parameterize the audio bitrate.

Reply:
* add tests.